mobile/examples/object_detections/iOS/ORTObjectDetection/ModelHandler.swift

```diff
@@ -0,0 +1,463 @@
+struct Inference {
+    let score: Float
+    let className: String
+    let rect: CGRect
+    let displayColor: UIColor
+}
+
+// Information about a model file or labels file.
+typealias FileInfo = (name: String, extension: String)
+
+enum OrtModelError : Error {
+    case error(_ message: String)
+}
+
+class ModelHandler: NSObject {
+    
+    // MARK: - Inference Properties
+    let threadCount: Int32
+    let threshold: Float = 0.5
+    let threadCountLimit = 10
+    
+    // MARK: - Model Parameters
+    let batchSize = 1
+    let inputChannels = 3
+    let inputWidth = 300
+    let inputHeight = 300
+    
+    private let colors = [
+        UIColor.red,
+        UIColor(displayP3Red: 90.0/255.0, green: 200.0/255.0, blue: 250.0/255.0, alpha: 1.0),
+        UIColor.green,
+        UIColor.orange,
+        UIColor.blue,
+        UIColor.purple,
+        UIColor.magenta,
+        UIColor.yellow,
+        UIColor.cyan,
+        UIColor.brown
+    ]
+    
+    private var labels: [String] = []
+    
+    init?(threadCount: Int32 = 1) {
+
+        self.threadCount = threadCount
+        
+        super.init()
+    }
+
+    // This methods preprocess the image, runs the ort inferencesession and returns the inference result
+    func runModel(onFrame pixelBuffer: CVPixelBuffer, modelFileInfo: FileInfo, labelsFileInfo: FileInfo) throws -> Result? {
+        
+        let modelFilename = modelFileInfo.name
+        
+        labels = loadLabels(fileInfo: labelsFileInfo)
+        
+        guard let modelPath = Bundle.main.path(
+            forResource: modelFilename,
+            ofType: modelFileInfo.extension
+        ) else {
+            print("Failed to get model file path with name: \(modelFilename).")
+            return nil
+        }
+        
+        let sourcePixelFormat = CVPixelBufferGetPixelFormatType(pixelBuffer)
+        assert(sourcePixelFormat == kCVPixelFormatType_32ARGB ||
+                sourcePixelFormat == kCVPixelFormatType_32BGRA ||
+                sourcePixelFormat == kCVPixelFormatType_32RGBA)
+        
+        let imageChannels = 4
+        assert(imageChannels >= inputChannels)
+        
+        let imageWidth = CVPixelBufferGetWidth(pixelBuffer)
+        let imageHeight = CVPixelBufferGetHeight(pixelBuffer)
+        
+        // Preprocess the image
+        let scaledSize = CGSize(width: inputWidth, height: inputHeight)
+        guard let scaledPixelBuffer = preprocess(ofSize: scaledSize, pixelBuffer) else {
+            return nil
+        }
+        
+        // Start the ORT inference environment
+        let env = try ORTEnv(loggingLevel: ORTLoggingLevel.warning)
+        let options = try ORTSessionOptions()
+        try options.setLogSeverityLevel(ORTLoggingLevel.verbose)
+        try options.setIntraOpNumThreads(self.threadCount) // TODO: check if calling the right methods
+        
+        let session = try ORTSession(env: env, modelPath: modelPath, sessionOptions: options)
```

Comment:
Move these into init() and make session a class variable?
Such that you don't need to create a session for every frame?

Reply:
I think this is a good idea. Looks like maybe this comment was resolved before it got addressed?

Reply:
More than a good idea - we don't want people copying this and creating a new session for each run as that's hugely expensive and will result in horrible performance.

Reply:
Sorry the comments just got missed. Also noticed couple typos. Will update together in another pr.

Reply:
After moving the session to initialization- init(), the `session.run()`will fail. Spent some time looking at it but couldn't figure out the reason. The ORTSession and modelhandler initializations are successul. Error occurs whenever we call `session.run(withInputs, outputNames, runoptions...)`...

The changes:
[https://github.com/microsoft/onnxruntime-inference-examples/compare/yguo/fix-ios-objdetect](https://github.com/microsoft/onnxruntime-inference-examples/compare/yguo/fix-ios-objdetect)

Appreciate any suggestions on potential reasons that could cause this error

Reply:
Try move the env to class variable also

Reply:
> Try move the env to class variable also

Thanks!! the app works and I updated.